### PR TITLE
gh-90300: Improve the Python CLI help output

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* document equivalent command-line options for all environment variables
* document equivalent environment variables for all command-line options
* reduce the size of variable and option descriptions to minimum
* sort the lists of -X options and variables that have equivalent command-line options
* remove the ending period in single-sentence descriptions

* Issue: gh-90300

----
📚 Documentation preview 📚: https://cpython-previews--115853.org.readthedocs.build/

### CODE REVIEW

**Comment by merwok** [comment]
I’m not sure if I agree with that, or if it’s the consensus within python-dev.
Could you post your message to the issue at least, or even discuss?

**Comment by serhiy-storchaka** [comment]
What can be done so that this PR does not hang for another 2 years like #93415? It is painful and errorprone to resolve merge conflicts after adding new options and environment variables.

**Comment by erlend-aasland** [comment]
There's always the possibility of opening a topic on Discourse, so we can have a broader discussion. I think I want to read up some more on the CLI guide Hugo linked to in the issue, before I decide.

**Comment by ericsnowcurrently** [comment]
FTR, this is a follow-up to gh-93415.

File: Python/initconfig.c [hunk 1]
```diff
@@ -173,8 +173,8 @@ Options (and corresponding environment variables):\n\
          also PYTHONWARNINGS=arg\n\
 -x     : skip first line of source, allowing use of non-Unix forms of #!cmd\n\
 -X opt : set implementation-specific option\n\
---check-hash-based-pycs always|default|never:\n\
-         control how Python invalidates hash-based .pyc files\n\
+--check-hash-based-pycs arg: control how Python invalidates hash-based .pyc\n\
+         files; arg is always|default|never\n\
```
Comment (hugovk) [inline Python/initconfig.c]: "argparse" style is more compact:
```suggestion
--check-hash-based-pycs {always|default|never}:\n
         control how Python invalidates hash-based .pyc files\n\
```

**Comment by serhiy-storchaka** [comment]
Opened a topic: https://discuss.python.org/t/and-again-about-the-python-cli-help/48401.

**Comment by hugovk** [comment]
It would be easier to review if the bullet points were in split into separate (followup) PRs. For example:

> * sort the lists of -X options and variables that have equivalent command-line options
> * remove the ending period in single-sentence descriptions

I think I could review and approve those as a single PR immediately, and it'd make followups smaller and easier (and quicker 🤞) to review.

**Comment by serhiy-storchaka** [comment]
The part of the problem is that not for all options and variables the equivalent is currently specified, and that there are two lists for environment variables. So the change may require adding references to options and variables (which may increase the size of the description, and may not fit the style of the verbose description) and moving variables from one list to other. Is it okay?

**Comment by serhiy-storchaka** [comment]
See #116739 for a simple step.

**Comment by serhiy-storchaka** [comment]
Next step:  #116756.

**Comment by serhiy-storchaka** [comment]
Next step: #116765.

File: Python/initconfig.c [hunk 2]
```diff
@@ -280,39 +254,27 @@ static const char usage_envvars[] =
...
+"PYTHONINTMAXSTRDIGITS: limit the size of int<->str conversions;\n"
+"                  0 disables the limit (-X int_max_str_digits=number)\n"
+"PYTHONNODEBUGRANGES: don't include extra location information in code objects\n"
+"                  (-X no_debug_ranges)\n"
 "PYTHONNOUSERSITE: disable user site directory (-s)\n"
 "PYTHONOPTIMIZE  : enable level 1 optimizations (-O)\n"
 "PYTHONPERFSUPPORT: support the Linux \"perf\" profiler (-X perf)\n"
 #ifdef Py_DEBUG
-"PYTHON_PRESITE=pkg.mod: import this module before site.py is run (-X presite)\n"
+"PYTHON_PRESITE: import this module before site.py is run (-X presite)\n"
```
Comment (merwok) [inline Python/initconfig.c]:
```suggestion
"PYTHON_PRESITE: import this module before site is imported (-X presite)\n"
```

File: Python/initconfig.c [hunk 3]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
+-X importtime: show how long each import takes; also PYTHONPROFILEIMPORTTIME\n\
+-X int_max_str_digits=number: limit the size of int<->str conversions;\n\
+         0 disables the limit; also PYTHONINTMAXSTRDIGITS\n\
+-X no_debug_ranges: don't include extra location information in code objects;\n\
+         also PYTHONNODEBUGRANGES\n\
+-X perf: support the Linux \"perf\" profiler; also PYTHONPERFSUPPORT.\n\
 "
 #ifdef Py_DEBUG
-"-X presite=package.module: import this module before site.py is run.\n"
+"-X presite=MOD: import this module before site.py is run; also PYTHON_PRESITE\n"
```
Comment (merwok) [inline Python/initconfig.c]:
```suggestion
"-X presite=MOD: import this module before site is imported; also PYTHON_PRESITE\n"
```

File: Python/initconfig.c (hunk 1 above)
Comment (merwok) [inline Python/initconfig.c]: The previous format seemed clear and in the same style as other switches with params

Comment (serhiy-storchaka) [inline Python/initconfig.c]:
```suggestion
--check-hash-based-pycs always|default|never:\n\
         control how Python invalidates hash-based .pyc files\n\
```

File: Python/initconfig.c [hunk 4]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
-         every instruction in code objects.  This is useful when smaller code\n\
-         objects and pyc files are desired as well as suppressing the extra\n\
-         visual location indicators when the interpreter displays tracebacks.\n\
--X perf: activate support for the Linux \"perf\" profiler by activating the\n\
-         \"perf\" trampoline.  When this option is activated, the Linux \"perf\"\n\
-         profiler will be able to report Python calls.  This option is only\n\
-         available on some platforms and will do nothing if is not supported\n\
-         on the current system.  The default value is \"off\".\n\
+         The default is \"on\" for installed Python and \"off\" for a local build.\n\
+         Also PYTHON_FROZEN_MODULES.\n\
```
Comment (hugovk) [inline Python/initconfig.c]: Replace full stops with semicolons like elsewhere?

File: Python/initconfig.c [hunk 5]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
+-X importtime: show how long each import takes; also PYTHONPROFILEIMPORTTIME\n\
+-X int_max_str_digits=number: limit the size of int<->str conversions;\n\
```
Comment (hugovk) [inline Python/initconfig.c]: `=number` or `=N`? There's a few `=N` in other places

```suggestion
-X int_max_str_digits=N: limit the size of int<->str conversions;\n\
```

File: Python/initconfig.c [hunk 6]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
+         0 disables the limit; also PYTHONINTMAXSTRDIGITS\n\
+-X no_debug_ranges: don't include extra location information in code objects;\n\
+         also PYTHONNODEBUGRANGES\n\
+-X perf: support the Linux \"perf\" profiler; also PYTHONPERFSUPPORT=1\n\
 "
 #ifdef Py_DEBUG
-"-X presite=package.module: import this module before site.py is run.\n"
+"-X presite=MOD: import this module before site is imported; also PYTHON_PRESITE\n"
```
Comment (hugovk) [inline Python/initconfig.c]: More concise:

```suggestion
"-X presite=MOD: import this module before site; also PYTHON_PRESITE\n"
```

File: Python/initconfig.c [hunk 7]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
-         tracing with a traceback limit of NFRAME frames\n\
--X utf8: enable UTF-8 mode for operating system interfaces, overriding the\n\
-         default locale-aware mode.  -X utf8=0 explicitly disables UTF-8 mode\n\
-         (even when it would otherwise activate automatically)\n\
--X warn_default_encoding: enable opt-in EncodingWarning for 'encoding=None'\
-"
-;
+         the interactive interpreter; only works on debug builds\n\
+-X tracemalloc[=N]: trace Python memory allocations; N sets a traceback limit\n\
+         of N frames (1 by default); also PYTHONTRACEMALLOC=N\n\
```
Comment (hugovk) [inline Python/initconfig.c]:
```suggestion
         of N frames (default: 1); also PYTHONTRACEMALLOC=N\n\
```

File: Python/initconfig.c [hunk 8]
```diff
@@ -280,39 +254,27 @@ static const char usage_envvars[] =
 "PYTHON_HISTORY  : the location of a .python_history file.\n"
 "\n"
 "These variables have equivalent command-line options (see --help for details):\n"
-"PYTHON_CPU_COUNT: Overrides the return value of os.process_cpu_count(),\n"
-"                  os.cpu_count(), and multiprocessing.cpu_count() if set to\n"
-"                  a positive integer. (-X cpu_count)\n"
+"PYTHON_CPU_COUNT: override the return value of os.cpu_count() (-X cpu_count)\n"
 "PYTHONDEBUG     : enable parser debug mode (-d)\n"
-"PYTHONDEVMODE   : enable the development mode (-X dev)\n"
+"PYTHONDEVMODE   : enable the Python development mode (-X dev)\n"
```
Comment (hugovk) [inline Python/initconfig.c]: Capitalised at https://docs.python.org/3/library/devmode.html, can skip "the"
```suggestion
"PYTHONDEVMODE   : enable Python Development Mode (-X dev)\n"
```

File: Python/initconfig.c [hunk 9]
```diff
@@ -280,39 +254,27 @@ static const char usage_envvars[] =
...
 "PYTHONDONTWRITEBYTECODE: don't write .pyc files (-B)\n"
 "PYTHONFAULTHANDLER: dump the Python traceback on fatal errors (-X faulthandler)\n"
-"PYTHON_FROZEN_MODULES: if this variable is set, it determines whether or not\n"
-"                  frozen modules should be used.  The default is \"on\" (or\n"
-"                  \"off\" if you are running a local build).\n"
-"                  (-X frozen_modules)\n"
+"PYTHON_FROZEN_MODULES: whether or not frozen modules should be used.\n"
```
Comment (hugovk) [inline Python/initconfig.c]: "or not" is redundant, remove full stop, make more direct/concise:

```suggestion
"PYTHON_FROZEN_MODULES: whether to use frozen modules\n"
```

File: Python/initconfig.c [hunk 10]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
-          * Enable asyncio debug mode\n\
-          * Set the dev_mode attribute of sys.flags to True\n\
-          * io.IOBase destructor logs close() exceptions\n\
--X faulthandler: enable faulthandler\n\
+-X cpu_count=N: override the return value of os.cpu_count();\n\
+         -X cpu_count=default cancels overriding; also PYTHON_CPU_COUNT\n\
+-X dev : enable Python development mode; also PYTHONDEVMODE\n\
+-X faulthandler: dump the Python traceback on fatal errors;\n\
+         also PYTHONFAULTHANDLER\n\
 -X frozen_modules=[on|off]: whether or not frozen modules should be used.\n\
```
Comment (hugovk) [inline Python/initconfig.c]:
```suggestion
-X frozen_modules=[on|off]: whether to use frozen modules\n\
```

File: Python/initconfig.c [hunk 11]
```diff
@@ -187,67 +187,41 @@ arg ...: arguments passed to program in sys.argv[1:]\n\
...
-            PyMem_SetupDebugHooks() C function\n\
-          * Enable the faulthandler module to dump the Python traceback on\n\
-            a crash\n\
-          * Enable asyncio debug mode\n\
-          * Set the dev_mode attribute of sys.flags to True\n\
-          * io.IOBase destructor logs close() exceptions\n\
--X faulthandler: enable faulthandler\n\
+-X cpu_count=N: override the return value of os.cpu_count();\n\
+         -X cpu_count=default cancels overriding; also PYTHON_CPU_COUNT\n\
+-X dev : enable Python development mode; also PYTHONDEVMODE\n\
```
Comment (hugovk) [inline Python/initconfig.c]:
```suggestion
-X dev : enable Python Development Mode; also PYTHONDEVMODE\n\
```

File: Python/initconfig.c [hunk 12]
```diff
@@ -280,39 +254,27 @@ static const char usage_envvars[] =
...
+"PYTHONDEVMODE   : enable the Python development mode (-X dev)\n"
 "PYTHONDONTWRITEBYTECODE: don't write .pyc files (-B)\n"
 "PYTHONFAULTHANDLER: dump the Python traceback on fatal errors (-X faulthandler)\n"
-"PYTHON_FROZEN_MODULES: if this variable is set, it determines whether or not\n"
-"                  frozen modules should be used.  The default is \"on\" (or\n"
-"                  \"off\" if you are running a local build).\n"
-"                  (-X frozen_modules)\n"
+"PYTHON_FROZEN_MODULES: whether or not frozen modules should be used.\n"
+"                  The default is \"on\" for installed Python and \"off\" for\n"
+"                  a local build. (-X frozen_modules)\n"
```
Comment (hugovk) [inline Python/initconfig.c]:
```suggestion
"                  a local build (-X frozen_modules)\n"
```

File: Python/initconfig.c [hunk 13]
```diff
@@ -280,39 +254,27 @@ static const char usage_envvars[] =
...
+"PYTHONINTMAXSTRDIGITS: limit the size of int<->str conversions;\n"
+"                  0 disables the limit (-X int_max_str_digits=number)\n"
+"PYTHONNODEBUGRANGES: don't include extra location information in code objects\n"
+"                  (-X no_debug_ranges)\n"
 "PYTHONNOUSERSITE: disable user site directory (-s)\n"
 "PYTHONOPTIMIZE  : enable level 1 optimizations (-O)\n"
 "PYTHONPERFSUPPORT: support the Linux \"perf\" profiler (-X perf)\n"
 #ifdef Py_DEBUG
-"PYTHON_PRESITE=pkg.mod: import this module before site.py is run (-X presite)\n"
+"PYTHON_PRESITE: import this module before site is imported (-X presite)\n"
```
Comment (hugovk) [inline Python/initconfig.c]:
```suggestion
"PYTHON_PRESITE: import this module before site (-X presite)\n"
```

**Review by hugovk — COMMENTED** [review]
Thanks, much easier to review now :)

Looks good, a few suggestions.

**Review by hugovk — APPROVED** [review]

**Pull request merged**